### ISSUE TITLE
tf.metrics.Mean* metrics miscalculated

### ISSUE THREAD

**Issue by gsimko** [tensorflow/tensorflow#35439]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): no
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): mac os 
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): pip3
- TensorFlow version (use command below): 2.0.0
- Python version: 3.7.4
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
tf.metrics.MeanAbsoluteError and others compute the mean of means.

**Describe the expected behavior**
They should compute the mean of all the data to support iterating over evaluation datasets.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

import tensorflow as tf
m=tf.metrics.MeanAbsoluteError()
m.update_state(y_true=[0,0], y_pred=[1,1])
m.update_state(y_true=[0], y_pred=[2])
assert m.result() == 2

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

The fix is simple, instead of the mean_squared_error function use a (new) sum_squared_error function in the metric and pass that to MeanMetricWrapper.

**Comment by ashutoshbsathe** [comment]
I think the behavior is expected since `update_state` accumulates ([\[1\]](https://github.com/tensorflow/tensorflow/blob/1cf0898dd4331baf93fe77205550f2c2e6c90ee5/tensorflow/python/keras/metrics.py#L213), [\[2\]](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/MeanAbsoluteError?version=stable#update_state)) the errors. You can verify this by checking the value of `m.total` variable after executing each `update_state` statement. You can do so by running following code:
```
import tensorflow as tf
m=tf.metrics.MeanAbsoluteError()
m.update_state(y_true=[0,0], y_pred=[1,1])
print(m.total.numpy())
m.update_state(y_true=[0], y_pred=[2])
print(m.total.numpy())
print(m.result().numpy())
```
Then since state count is 2 and total is 3, the result returned is 1.5. You can also refer to the source code of `result()` [here](https://github.com/tensorflow/tensorflow/blob/r2.0/tensorflow/python/keras/metrics.py#L361-L371).

**Comment by gsimko** [comment]
Of course the result matches the source code :) However, I believe the source code does not match the intent.

If you do an eval loop like this where the data is sequential hence label is (batch,step,1)-shaped, it won't calculate the MAE/MSE properly:
for feature, label in eval_dataset:
   pred = model.predict(feature)
   metric.update_state(pred, label)

The reason is that "mean of means of squares" != "mean of squares" as long as in the first term the dimension varies.

**Comment by ashutoshbsathe** [comment]
Oh now I got it. Sorry for misinterpreting your question the first time. @gsimko

**Comment by ymodak** [comment]
To avoid metric accumulation you may try `reset_states()` method.
```python
import tensorflow as tf
m=tf.metrics.MeanAbsoluteError()
m.update_state(y_true=[0,0], y_pred=[1,1])
m.reset_states()
m.update_state(y_true=[0], y_pred=[2])
assert m.result() == 2
```

**Comment by gsimko** [comment]
Resetting the state does not solve the problem - during evals one needs to accumulate the metrics. Why was this issue closed?

**Comment by pavithrasv** [comment]
@gsimko Thank you for the question. For metrics and losses, we expect that labels and predictions are at least 2D.

https://www.tensorflow.org/api_docs/python/tf/keras/metrics/MeanAbsoluteError?version=nightly#update_state
```
y_true: Ground truth values. shape = `[batch_size, d0, .. dN]`, except
        sparse loss functions such as sparse categorical crossentropy where
        shape = `[batch_size, d0, .. dN-1]`
y_pred: The predicted values. shape = `[batch_size, d0, .. dN]`
```

For example, let's say y_pred and y_true are of the shape [batch_size, d0], we compute the mean across the last axis in a sample -> this will give us one value per-sample -> `[batch_size]`.  The second mean is computed across all samples across all batches.

**Comment by tensorflowbutler** [comment]
It has been 19 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by gsimko** [comment]
@pavithrasv: thanks for the response, that answered it! I also just realized that my original assertion was off (it should have been 4/3 instead of 2), and by following what you described (adding an inner dimension) I get that answer.

If someone else finds this problem, this works as expected:
import tensorflow as tf
m=tf.metrics.MeanAbsoluteError()
m.update_state(y_true=[[0],[0]], y_pred=[[1],[1]])
m.update_state(y_true=[[0]], y_pred=[[2]])
assert (m.result() - 4/3)<1e-6
